src/service/locator.cpp

```diff
@@ -280,19 +280,50 @@ locator_cfg_t::locator_cfg_t(const std::string& name_, const dynamic_t& root):
     restricted.insert(name);
 }
 
+results::routing
+locator_t::routing_t::dump() const {
+    auto results = results::routing();
+    auto builder = std::inserter(results, results.end());
+
+    auto mapping = groups.synchronize();
+
+    std::transform(mapping->begin(), mapping->end(), builder,
+        [](const rg_map_t::value_type& value) -> results::routing::value_type
+    {
+        return {value.first, value.second.all()};
+    });
+
+    return results;
+}
+
+streamed<results::routing>
+locator_t::routing_t::notify(const std::string& ruid, const results::routing& results, policy_t policy) {
+    auto stream = routers.apply([&](router_map_t& mapping) -> streamed<results::routing> {
+        if(mapping.count(ruid) == 0 || ((policy == policy_t::replace) && mapping.erase(ruid))) {
```

Comment:
in case of replace we do 2 searches in map. It's not critical, but we can use find here and erase by iterator.